### PR TITLE
Implement Lab1

### LINKED ISSUE #147: Lab 1 - fibonacci series
Create an executable go program in directory 01_fib/brvishnu92.
Write a function that prints the first n Fibonacci numbers.
Write test cases covering all function runs.

### PR DESCRIPTION
**Pull request by brvishnu92**
Fixes #147

Review of colleague's PR #179

Changes proposed in this PR:
- Implement Lab1 Fibonacci and negafibonacci.
- Test Lab1

### CODE REVIEW

**Review by jonathanheaden — DISMISSED** [review]
the convention in go is to use very short variable names inside a code block. The idea is that you will be faster to parse these while you are in a small context of the code block. a variables like `isPositive` whilst readable is longer than it needs to be in this context. It is an important style to adapt to because the rest of the go community will use short var names. I would also suggest refactoring `first`,`second` etc 
Apart from this everything else looks good to me

**Comment by brvishnu92** [comment]
@jonathanheaden Thanks for the feedback and I will implement these standards. I have modified the code as well. Please review it.

File: 01_fib/brvishnu92/main_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+5
+-8
+13
+`)},
+		"0": {input: 0, expected: strconv.Quote(`0
+`)},
+	}
+
+	for name, test := range testCases {
+		test := test
```
Comment (eastrd) [inline 01_fib/brvishnu92/main_test.go:56]: Line 56 seems extra, could you remove it?

File: 01_fib/brvishnu92/main_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+8
+13
+`)
+	actual := strconv.Quote(buf.String())
+	if expected != actual {
+		t.Errorf("Expected %v, got %v ", actual, expected)
+	}
+}
+
+func TestMainFib(t *testing.T) {
```
Comment (eastrd) [inline 01_fib/brvishnu92/main_test.go]: Consider rename it to TestFib as you are only testing fib function.

**Review by eastrd — CHANGES_REQUESTED** [review]

File: 01_fib/brvishnu92/main_test.go (hunk 1 above)
Comment (brvishnu92) [reply 01_fib/brvishnu92/main_test.go:56]: @eastrd : If I don't do that assignment it comes up with the following linting error :
Using the variable on range scope `test` in function literal.

File: 01_fib/brvishnu92/main_test.go (hunk 2 above)
Comment (brvishnu92) [reply 01_fib/brvishnu92/main_test.go]: Oh yeah, thanks for that. I have updated it!!

**Comment by brvishnu92** [comment]
Git rebase isnt showing my previous commits. :-(

Its showing noop in place of commits when i do git rebase -i

**Review by eastrd — DISMISSED** [review]
The codes look good to me, sorry I cannot help you with rebase issues.

**Review by eastrd — DISMISSED** [review]

File: 01_fib/brvishnu92/main.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	case n == 0:
+		fmt.Fprintln(out, 0)
+	case n < 0:
+		isPos = false
+	}
+
+	n1, count := 1, 1
+	n2, sum := 0, 0
+	n = int(math.Abs(float64(n)))
+	for count <= n {
```
Comment (anzboi) [inline 01_fib/brvishnu92/main.go]: You don't need the count variable, you can use a c-style for loop.
```
for i := 0; i < n; i++
```

File: 01_fib/brvishnu92/main.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		sum = n1 + n2
+		if !isPos && count%2 == 0 {
+			fmt.Fprintln(out, sum*-1)
+		} else {
+			fmt.Fprintln(out, sum)
+		}
+		n1, n2 = n2, sum
```
Comment (anzboi) [inline 01_fib/brvishnu92/main.go]: Strictly speaking, you don't need to keep track of `sum`, though this is more readable so I'll leave it up to you if you want to refactor or not.

File: 01_fib/brvishnu92/main_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+`)},
+		"negative": {input: -7, expected: strconv.Quote(`1
+-1
+2
+-3
+5
+-8
+13
+`)},
+		"0": {input: 0, expected: strconv.Quote(`0
```
Comment (anzboi) [inline 01_fib/brvishnu92/main_test.go:51]: 👍 for the edge case

**Review by anzboi — CHANGES_REQUESTED** [review]
This is good to go, however you PR deletes @kasunfdo 's lab 1, can you fix your branch so that it is not deleted.
I've added a couple of other comments you can fix while you're at it if you want. I've also fixed the formatting in your PR message at the top.

File: 01_fib/brvishnu92/main.go (hunk 4 above)
Comment (brvishnu92) [reply 01_fib/brvishnu92/main.go]: Oh yeah, thanks to you I found a cleaner way to implement :-)

File: 01_fib/brvishnu92/main.go (hunk 3 above)
Comment (brvishnu92) [reply 01_fib/brvishnu92/main.go]: @anzboi Sure, I have implemented these changes 👍

**Review by juliaogris — DISMISSED** [review]
Code looks great. 
Only thing left: rework your commit history so that Kasuns lab work is untouched in this PR.
Try it on a throw away branch first.
Full disclosure: I haven't tried this locally myself - danger/possible typos ahead ;)
@camh-anz is our resident git master, he might have something to add/correct.
```bash
git checkout master
git pull upstream master
git checkout -b lab1-test
git cherrypick 820fdb7229f1b7493e3950b7727611adc955debe
git cherrypick d7c1959a113ec8bdbc75ac746f1f9ad66cbe9530
# if you are happy with the result: (testing, linting still ok, `git show` gives you good commit messages/changes)
git branch -m lab1 lab1-old
git branch -m lab1
git push -f
```

**Comment by camscale** [comment]
rather than cherrypick, I'd just rebase the two commits onto an updated master:
```
git checkout master
git pull upstream master
git rebase --onto master lab1~2 lab1
# This will leave you on the lab1 branch. Check the result (testing, linting, etc)
git push -f
```

That will rebase the last two commits of lab1 (`lab1~2`) onto master.

**Review by anzboi — APPROVED** [review]
LGTM 👍 
@juliaogris to approve now

**Comment by brvishnu92** [comment]
@juliaogris Can you please approve changes? I am unable to merge. Thanks :-)

**Comment by brvishnu92** [comment]
@anzboi Can you please help me to merge?  I spoke to Julia and she suggested to remove her review. Thanks.

**Pull request merged**